### ISSUE TITLE
uuids, or something similar, for collections?

### ISSUE THREAD

**Issue by NoisyFlowers** [azgs/azlibrary_database#11]
Andrew's use of uuids in the azgs_glossary has got me thinking. 

I've been bothered for a while by the fact that the name of an archived collection is tied to our internal primary key for that collection. Those primary key values can change for any number of reasons. Similarly, it's been bothering me that we expose our internal primary keys through the api (as the metadata route is currently implemented).

Also, the only natural key we have for each collection is the azgs_old_url, but that only works because of the old id numbers for the 1348 we currently have in there. What about the mining collections? Or other collections to come?

Perhaps each collection needs a permanent unique identifier.

For various reasons, I don't want to use a uuid for the internal primary keys on any tables. But, I'm wondering about generating one for each collection that we use as this permanent identifier. This id would be used in naming the archive file and in urls that access that collection.

I'm putting this as High Priority because it's a decision we should make sooner than later.

**Comment by NoisyFlowers** [comment]
To be more clear: the uuid would not be used as primary key for the collections table. The use of a uuid would decouple identification of collections outside the db from inside.

**Comment by aazaff** [comment]
So the collection_id field would remain the PKEY and be an integer sequence, but there would be another field called uuid (or whatever) - kind of similar to how I had the azgs_glossary tale organized?

And this uuid field would be what you use for naming the folders etc? What would go in the azgs.json collection_id field, the internal PKEY or the new UUID?

**Comment by NoisyFlowers** [comment]
I haven't thought that through yet. My gut says to store the uuid, let's call it perm_id for now, but I have to think through the ramifications of that. 

I guess the first question I have is whether you think my concerns have merit and if it's worth considering this rather significant change.

**Comment by aazaff** [comment]
I support it because I honestly hate having the .tar folders just having integer names... to easy for there to be some kind of weird accidental overwrite/pointer error.

Is it really that significant of a change though?

**Comment by NoisyFlowers** [comment]
I think I already have this implemented. Public.collections has a new column called perm_id. This id is what is used to name the archive tarball. It is also stored in azgs.json in that tarball. It is what is used to determine if this is a new collection or an update to an old one. Thus, this id will survive db rebuilds. 

I'm leaving collection_id in the azgs.json as well for now, though it is effectively unused. 

Before I commit this, I want to make sure you're ok with tarball archives that look like this: a6b0b97c-3930-11e9-84f3-080027ff4385.tar.gz. I suppose it's no more cryptic than 1.tar.gz. Just longer.

This has also left me wondering about the collection_group id that we currently store in azgs.json. It's just as likely that could change in the future. Should we consider doing this for that table as well? Or maybe just storing the group name in the azgs.json and omitting the id (assuming we put a unique constraint on that column in public.collection_groups.

**Comment by aazaff** [comment]
1. Can we simplify the hash? You're right it's not any more cryptic than the collection_id integers, but just seems like overkill.... What if we abbreviated the collection_group_names and used that as some sort of prefix?

2. I would say keep collection_group_name in the azgs.json and just omit the collection_group_id field.

**Comment by NoisyFlowers** [comment]
Regarding (1): I had only one goal in proposing this: to create collections ids that could reliably survive db rebuilds. Using UUID was quick to implement, and foremost in my mind, since we are already using them in the glossary. 

However, I'm seeing now that you may have some other ideal constraints we'd want permanent id's to meet (like perhaps a readable portion, shorter than 128 bits). Can you list these?

Perhaps I can build some other sort of generator. The thing is, UUID's are overkill but reliably unique. I thought about using epoch millis, but there is a chance of collision anytime the system updates from a central timing authority.

**Comment by aazaff** [comment]
I am leaning towards some kind of collection_group based abbreviation followed by a sequential integer or a randomised hash?

Digital Geologic Map becomes

DGM111138 or something like that? I can think of a lot of benefits of this. We could specify the abbreviation as a column in the collection_groups table.

How do you feel about this?

**Comment by NoisyFlowers** [comment]
I think this could work. I wouldn't use a serial generator because that would lead to collisions when the db is rebuilt.  Maybe epoch millis with a random number appended. That would minimize the chance of collision on system time correction. 

When can you generate the collection_group abbreviations?

**Comment by aazaff** [comment]
Immediately, I will put them in the tmplibrary table?

**Comment by NoisyFlowers** [comment]
Or you could just send them to me. Whatever is easiest. I'll have to add them to the schema creation sql

**Comment by NoisyFlowers** [comment]
Ok, I've implemented this. Had to use a trigger on the perm_id column rather than a default because the function needs to access the collection_group_id of the new row. 

The new id format looks like this: ADGM-1551136697550-107. I bracketed the time stamp with dashes to make it easy to split it off if we need to for some reason. The last bit is just a random number I append to minimize the possibility of time-related collisions. To be honest, it's probably not necessary, but I feel better with it there. 

I'll test this some more, then put it on dev tomorrow and test it further there.
